### PR TITLE
Initial NET5/ILLink support

### PR DESCRIPTION
**Pull request by radekdoulik**
Implements https://github.com/xamarin/xamarin-android/issues/4707

With .NET5 we are switching from XA linker implemented as msbuild task to ILLink tool with [custom steps](https://github.com/mono/linker/blob/master/docs/custom-steps.md) provided in an assembly.

The new assembly with our custom steps is named `Microsoft.Android.Sdk.ILLink` and the source code is located in `src/Microsoft.Android.Sdk.ILLink`. It is used during build by `ILLink` tool.

The initial support is already able to link and run simple XA and XA/XF samples.

Future work
* replace reflection usage in SetupStep https://github.com/xamarin/xamarin-android/issues/4709
* audit reflection usage in our code https://github.com/xamarin/xamarin-android/issues/4708
* add more customs steps - we are still missing MonoDroidMarkStep

Notes
* the Profile API is not available in the linker [public API](https://github.com/mono/linker/tree/master/src/linker/ref). we work around it by adding `src/Microsoft.Android.Sdk.ILLink/Profile.cs` with missing pieces.
* `TypeDefinition.GetMethods` extension method from `Mono.Tuner.MethodBodyRocks` is not public. It is very simple so we are inlining it.
* the ILLink is now enabled in `Release` configuration defaults.
* tests were updated to reference more packages, to not miss second level dependencies.

Initial results:

apkdiff output summary for HelloAndroid sample, comparing `Debug` and `Release` apk's
```
  -       7,300 Davik executables -2.34% (of 311,436)
  -  14,733,944 Assemblies -31.85% (of 46,265,064)
  -   2,492,452 Shared libraries -11.43% (of 21,803,164)
  -  15,528,456 Package size difference -28.70% (of 54,114,462)
```

Context:
* `_RunILLink` target https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets
* linker documentation https://github.com/mono/linker/tree/master/docs

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.props [hunk 1]
```diff
@@ -39,9 +39,12 @@
     <AndroidUseSharedRuntime Condition=" '$(AndroidUseSharedRuntime)' == '' ">false</AndroidUseSharedRuntime>
     <EmbedAssembliesIntoApk Condition=" '$(EmbedAssembliesIntoApk)' == '' ">true</EmbedAssembliesIntoApk>
     <!--FIXME: Disable linking by default -->
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.props:41]: You can probably take out this comment.

File: src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs [hunk 2]
```diff
@@ -17,13 +17,12 @@
...
 namespace Xamarin.Android.Tasks
 {
-	public class MonoAndroidHelper
+	public partial class MonoAndroidHelper
 	{
 		static Lazy<string> uname = new Lazy<string> (GetOSBinDirName, System.Threading.LazyThreadSafetyMode.PublicationOnly);
 
 		// Set in ResolveSdks.Execute();
 		// Requires that ResolveSdks.Execute() run before anything else
-		public static string[] TargetFrameworkDirectories;
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:26]: Does anything set this when the linker is running? I would guess it lives in a different `static` scope than the rest of XA's build?

File: src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.Linker.targets [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+This file contains the .NET 5-specific targets to customize ILLink
+
+***********************************************************************************************
+-->
+
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <Target Name="_PrepareLinking" AfterTargets="ComputeResolvedFilesToPublishList">
+    <Message Text="rodo: prepare linking" Importance="high" />
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.Linker.targets:13]: Maybe temporary message?

File: src/Xamarin.Android.Linker/SetupStep.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+{
+	class SetupStep : BaseStep
+	{
+		List<IStep> _steps;
+
+		List<IStep> Steps {
+			get {
+				if (_steps == null) {
+					var pipeline = typeof (LinkContext).GetProperty ("Pipeline").GetGetMethod ().Invoke (Context, null);
+					_steps = (List<IStep>) pipeline.GetType ().GetField ("_steps", BindingFlags.Instance | BindingFlags.NonPublic).GetValue (pipeline);
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Linker/SetupStep.cs:20]: Should we tell Marek Safar what you had to do here? Do we need some new public property here?

Comment (radekdoulik) [reply src/Xamarin.Android.Linker/SetupStep.cs:20]: I am going to add it to the "ILLinker Migration to .NET 5 APIs" document.

File: src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs (hunk 2 above)
Comment (radekdoulik) [reply src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:26]: Nothing sets it when used in ILLlink.

It is used through `IsFrameworkAssembly` in `MarkJavaObjects`. Maybe we can replace the `IsFrameworkAssembly` with `IsSdk`, `IsProduct`  calls?

File: src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs [hunk 5]
```diff
@@ -289,16 +300,24 @@ void AddNewExceptionMethod (TypeDefinition type, MethodDefinition method)
 
 		public virtual void LogMessage (string message)
 		{
+#if !NET5_LINKER
 			Context.LogMessage (message);
+#else
+			Context.LogMessage (MessageContainer.CreateInfoMessage (message));
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs:306]: Could we introduce an extension method which provides a `LogMessage(this LinkerContext context, string message)` extension. We could then remove a number of these `#if` blocks in the code.

Comment (radekdoulik) [reply src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs]: good idea 👍

File: src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs (hunk 2 above)
Comment (radekdoulik) [reply src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs:26]: In the end I set it in the setup step, passing the paths through the `--custom-data`

**Comment by radekdoulik** [comment]
> Do you know what .NET 5 version we need for this to work?

Locally I am using 5.0.100-preview.5.20229.12. I will try to bump it to lastest.

**Comment by radekdoulik** [comment]
Indeed. dotnet tests are now failing with related error:
```
error XARSD7004: System.ArgumentException: `C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0-preview.6.20264.1\ref\netcoreapp5.0` must be a directory! (Parameter 'frameworkDirectories')
```

**Comment by jonpryor** [comment]
[`DotNetBuild("android.21-arm",True)` on Windows fails](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3740519&view=ms.vss-test-web.build-test-results-tab&runId=13306362&resultId=100086&paneView=attachments):

```
The assembly 'C:\A\vs2019xam00000L-1\_work\2\s\packages\microsoft.android.sdk\10.0.100-ci.pr.gh4669.53\targets\..\tools\Xamarin.Android.Linker.dll' specified for '--custom-step' option could not be found (TaskId:185)
18:23:35.119   1:7>C:\Program Files\dotnet\sdk\5.0.100-preview.6.20265.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ILLink.targets(83,5): error MSB6006: "dotnet.exe" exited with code 1. [C:\A\vs2019xam00000L-1\_work\2\s\bin\TestRelease\temp\DotNetBuildandroid.21-armTrue\UnnamedProject.csproj]
                   Done executing task "ILLink" -- FAILED. (TaskId:185)
```

File: src/Xamarin.Android.Linker/SetupStep.cs (hunk 4 above)
Comment (marek-safar) [reply src/Xamarin.Android.Linker/SetupStep.cs:20]: This is an incorrect approach. All steps should set from msbuild target. We won't expose List<IStep> as that's an implementation detail which we are going to change

Comment (radekdoulik) [reply src/Xamarin.Android.Linker/SetupStep.cs:20]: Yes, I am aware of that. That is why I mentioned `Context.Pipeline` and `Pipeline` methods in the ILLinker-Migration-to-.NET-5-APIs document, to be considered for adding to the public API. This way we would not need reflection and access to the list itself.

Comment (radekdoulik) [reply src/Xamarin.Android.Linker/SetupStep.cs:20]: Discussed on Slack and we might not be able to have public Pipeline API. I would like to postpone the resolution for later. Opened https://github.com/xamarin/xamarin-android/issues/4709 for it. Will look into it in future PR.

File: src/Xamarin.Android.Linker/Xamarin.Android.Linker.csproj [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  <Import Project="..\..\Configuration.props" />
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <DefineConstants>NET5_LINKER</DefineConstants>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <OutputPath>$(XamarinAndroidSourcePath)bin\Build$(Configuration)\packs\tools</OutputPath>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.ILLink" Version="5.0.0-preview.3.20227.3" />
+    <ProjectReference Include="..\..\external\Java.Interop\src\Java.Interop.Export\Java.Interop.Export.csproj" />
```
Comment (jonpryor) [inline src/Xamarin.Android.Linker/Xamarin.Android.Linker.csproj:11]: I do not understand why we would have a `@(ProjectReference)` to `Java.Interop.Export.dll`.  I don't see any mentions of types located within that assembly within this PR, e.g. neither `JniSignature` nor `MarshalMemberBuilder` make an appearance.

File: Xamarin.Android.Build.Tasks.sln [hunk 7]
```diff
@@ -25,11 +25,13 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Mono.Debugging.Soft", "exte
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "vswhere", "tools\vswhere\vswhere.csproj", "{DBDC804F-8406-4F5E-83C6-720CB0CB6C6F}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Xamarin.Android.Tools.AndroidSdk", "external\xamarin-android-tools\src\Xamarin.Android.Tools.AndroidSdk\Xamarin.Android.Tools.AndroidSdk.csproj", "{E34BCFA0-CAA4-412C-AA1C-75DB8D67D157}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Xamarin.Android.Tools.AndroidSdk", "external\xamarin-android-tools\src\Xamarin.Android.Tools.AndroidSdk\Xamarin.Android.Tools.AndroidSdk.csproj", "{E34BCFA0-CAA4-412C-AA1C-75DB8D67D157}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Xamarin.Android.Linker", "src\Xamarin.Android.Linker\Xamarin.Android.Linker.csproj", "{2A237E71-1FA3-409B-B13E-F6294932A5A9}"
```
Comment (jonpryor) [inline Xamarin.Android.Build.Tasks.sln:30]: I half wonder if this should be `Microsoft.Android.Linker`. :-/

File: src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/MarkJavaObjects.cs [hunk 8]
```diff
@@ -165,7 +166,7 @@ bool CheckInvokerType (TypeDefinition type, string name)
 
 		void PreserveInterfaceMethods (TypeDefinition type, TypeDefinition invoker)
 		{
-			foreach (var m in type.GetMethods ()) {
+			foreach (var m in type.Methods.Where (m => !m.IsConstructor)) {
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/MarkJavaObjects.cs:169]: Any specific reason for this change?  The PR description is anemic.

File: Xamarin.Android.Build.Tasks.sln (hunk 7 above)
Comment (marek-safar) [reply Xamarin.Android.Build.Tasks.sln:30]: Or even better `Microsoft.Android.Sdk.ILLink` ;-)

Comment (jonpryor) [reply Xamarin.Android.Build.Tasks.sln:30]: How many assemblies do we want running around?  That's a rather fine-grained name…

File: src/Xamarin.Android.Linker/Xamarin.Android.Linker.csproj (hunk 6 above)
Comment (radekdoulik) [reply src/Xamarin.Android.Linker/Xamarin.Android.Linker.csproj:11]: This is to have `MarshalMemberBuilder.GetMarshalMethodName` for `TryGetMarshalMethod` method.

**Pull request merged**